### PR TITLE
new keep notifications using slack

### PR DESCRIPTION
**Pull request by camhashemi**
@andrewconner 

you mentioned putting the attribution info in the `extras` object, not sure how that's used since all of the relevant info is in the rest of the object, but I see we add `keep.url` even though that's in the `url` field already.

I don't remember if you ever showed me how to run the ext on a local server, but that'd be good for this.

### CODE REVIEW

File: kifi-backend/modules/eliza/app/com/keepit/notify/info/NotificationKindInfoRequests.scala [hunk 1]
```diff
@@ -9,6 +9,10 @@ import com.keepit.notify.info.NotificationInfoRequest._
 import com.keepit.notify.model.event._
 import play.api.libs.json.Json
 
+object ExtensionImageUrls {
+  val SLACK: String = "/images/slack-245.png"
```
Comment (andrewconner) [inline kifi-backend/modules/eliza/app/com/keepit/notify/info/NotificationKindInfoRequests.scala]: Ah I didn't know the use case when you asked about this. Let's stick it on the CDN. Here's a 200x200:

https://djty7jcqog9qu.cloudfront.net/oa/98c4c6dc6bf8aeca952d2316df5b242b_200x200-0x0-200x200_cs.png

File: kifi-backend/modules/shoebox/angular/img/svg/slack.svg [hunk 2]
```diff
@@ -1,4 +1,3 @@
-<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/angular/img/svg/slack.svg:1]: On purpose?

File: packrat/images/results/slack.svg [hunk 3]
```diff
@@ -0,0 +1,12 @@
+<svg viewBox="0 0 256 256" preserveAspectRatio="xMidYMid" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
```
Comment (andrewconner) [inline packrat/images/results/slack.svg]: Is this used?

**Comment by andrewconner** [comment]
:+1: Great. Does this cover the case where we know who the Slack user is? (ie, me, in Slack, post a link as @ac, and Kifi knows that's me as a user)

File: kifi-backend/modules/eliza/app/com/keepit/notify/info/NotificationKindInfoRequests.scala (hunk 1 above)
Comment (camhashemi) [inline kifi-backend/modules/eliza/app/com/keepit/notify/info/NotificationKindInfoRequests.scala]: thanks, I don't know the best way to do these things

File: kifi-backend/modules/eliza/app/com/keepit/notify/info/NotificationKindInfoRequests.scala [hunk 4]
```diff
@@ -100,11 +104,22 @@ class NotificationKindInfoRequests @Inject()(implicit val pubIdConfig: PublicIdC
       val newKeep = RequestKeep(event.keepId).lookup(batched)
       val keeper = RequestUserExternal(newKeep.ownerId).lookup(batched)
       val libraryKept = RequestLibrary(event.libraryId).lookup(batched)
+
+      val attributionOpt = newKeep.attribution
+      val slackImage: Option[NotificationImage] = attributionOpt.map { case (_, userOpt) =>
+        userOpt.map(UserImage).getOrElse(PublicImage(ExtensionImageUrls.SLACK))
```
Comment (camhashemi) [inline kifi-backend/modules/eliza/app/com/keepit/notify/info/NotificationKindInfoRequests.scala]: @andrewconner if it's a slack keep, use the attributed user's image, else the slack image.

Comment (andrewconner) [inline kifi-backend/modules/eliza/app/com/keepit/notify/info/NotificationKindInfoRequests.scala]: Sweet

File: kifi-backend/modules/eliza/app/com/keepit/notify/info/NotificationKindInfoRequests.scala (hunk 1 above)
Comment (andrewconner) [inline kifi-backend/modules/eliza/app/com/keepit/notify/info/NotificationKindInfoRequests.scala]: I just resized it to be 200x200, then uploaded as an org avatar

File: kifi-backend/modules/eliza/app/com/keepit/notify/info/NotificationKindInfoRequests.scala [hunk 5]
```diff
@@ -112,7 +133,8 @@ class NotificationKindInfoRequests @Inject()(implicit val pubIdConfig: PublicIdC
           "library" -> Json.toJson(libraryKept),
           "keep" -> Json.obj(
             "id" -> newKeep.id,
-            "url" -> newKeep.url
+            "url" -> newKeep.url,
+            "attr" -> Json.toJson(newKeep.attribution)
```
Comment (camhashemi) [inline kifi-backend/modules/eliza/app/com/keepit/notify/info/NotificationKindInfoRequests.scala:137]: @andrewconner @jaredpetker @DerekBurch do I need to make these keep notifications backwards compatible? background: adding slack attribution to them

Comment (jaredpetker) [inline kifi-backend/modules/eliza/app/com/keepit/notify/info/NotificationKindInfoRequests.scala:137]: Should be fine for mobile, mobile never handled attributions before for push notifications anyways, we can add it, but to answer your question, there's nothing else on your ned to do for it.

File: kifi-backend/modules/eliza/app/com/keepit/notify/info/NotificationKindInfoRequests.scala [hunk 6]
```diff
@@ -9,6 +9,10 @@ import com.keepit.notify.info.NotificationInfoRequest._
 import com.keepit.notify.model.event._
 import play.api.libs.json.Json
 
+object ImageUrls {
+  val SLACK_LOGO: String = "https://djty7jcqog9qu.cloudfront.net/oa/98c4c6dc6bf8aeca952d2316df5b242b_200x200-0x0-200x200_cs.png"
```
Comment (camhashemi) [inline kifi-backend/modules/eliza/app/com/keepit/notify/info/NotificationKindInfoRequests.scala:13]: @andrewconner do you know a better place for this?

File: kifi-backend/modules/eliza/app/com/keepit/notify/info/NotificationKindInfoRequests.scala (hunk 5 above)
Comment (camhashemi) [inline kifi-backend/modules/eliza/app/com/keepit/notify/info/NotificationKindInfoRequests.scala:137]: :+1: perfect answer

File: kifi-backend/modules/eliza/app/com/keepit/notify/info/NotificationKindInfoRequests.scala (hunk 6 above)
Comment (andrewconner) [inline kifi-backend/modules/eliza/app/com/keepit/notify/info/NotificationKindInfoRequests.scala:13]: That's fine

**Comment by ryanpbrewster** [comment]
:boom:, nice work man

**Pull request merged**